### PR TITLE
[Update]: Pass word timing to PreviewCanvas on double click

### PR DESCRIPTION
**Pull request by allishultes**
**Is your Pull Request request related to another [issue](https://github.com/bbc/react-transcript-editor/issues) in this repository ?**      

[Issue #85 in DPE Client](https://github.com/bbc/digital-paper-edit-client/issues/85).

**Describe what the PR does**    

- [x] Passes a new prop for current word time to the PreviewCanvas component
- [x] Re-calculates start and end times of words in paper-cut to correspond to programme script + playlist
- [x] Un-comments the double-click functionality in the transcript container view 

**State whether the PR is ready for review or whether it needs extra work**    

Ready for review. To review please test in conjunction [with PR #53](https://github.com/bbc/digital-paper-edit-storybook/pull/53) in the DPE Storybook. 

**Additional context**    

To test, please **create a new paper-edit** in a pre-existing project and do the following to confirm they all work as expected:
- [x] Add new paper-cuts to the end of the programme script. Click on words in each paper-cut to make sure the video jumps to the correct timing and the PreviewCanvas renders appropriately
- [x] Move the insert point so you're adding new paper-cuts to midway through a paper-cut. Click on words in each paper-cut to make sure the video jumps to the correct timing and the PreviewCanvas renders appropriately
- [x] Re-arrange the programme script. Click on words in each paper-cut to make sure the video jumps to the correct timing and the PreviewCanvas renders appropriately
- [x] Delete paper-cuts from the programme script. Click on words in each paper-cut to make sure the video jumps to the correct timing and the PreviewCanvas renders appropriately
- [x] Double-click a word in the TranscriptTabContainer (left-hand side). This should also jump to the appropriate point in the transcript.

### CODE REVIEW

File: src/Components/PaperEditor/ProgrammeScriptContainer/index.js [hunk 1]
```diff
@@ -307,6 +299,7 @@ const ProgrammeScript = props => {
       // Video context probably needs more info like, which clip/track in the sequence?
       // investigate how to set currentTime in video context
```
Comment (emettely) [inline src/Components/PaperEditor/ProgrammeScriptContainer/index.js:300]: You can remove this comment too 👍 
```suggestion
      // investigate how to set currentTime in video context
```

**Comment by emettely** [comment]
I've tried this locally and I can't seem to get the video to auto play, and I can't see that the setting the `timeOnClick` state actually updates the PreviewCanvas, or even triggers a useEffect change. I might be missing something, let me know.

**Comment by allishultes** [comment]
> I've tried this locally and I can't seem to get the video to auto play, and I can't see that the setting the `timeOnClick` state actually updates the PreviewCanvas, or even triggers a useEffect change. I might be missing something, let me know.

We can look at it tomorrow - it is autoplaying and jumping to the right time for me locally, so not sure exactly why we have different results (or whether I've just misunderstood something).

File: src/Components/PaperEditor/ProgrammeScriptContainer/index.js [hunk 2]
```diff
@@ -35,6 +41,7 @@ const ProgrammeScriptContainer = (props) => {
   const [ elements, setElements ] = useState();
   const [ title, setTitle ] = useState('');
   const [ resetPreview, setResetPreview ] = useState(false);
+  const [ timeOnClick, setTimeOnClick ] = useState();
```
Comment (emettely) [inline src/Components/PaperEditor/ProgrammeScriptContainer/index.js:44]: I would call it just `currentTime` cause that's what you're assigning it to :)
```suggestion
  const [ currentTime, setCurrentTime ] = useState();
```

File: src/Components/PaperEditor/ProgrammeScriptContainer/index.js [hunk 3]
```diff
@@ -143,12 +149,17 @@ const ProgrammeScriptContainer = (props) => {
...
-    if (resetPreview && elements && elements.length > 0) {
+    const handleUpdatePreview = async () => {
+      const newPlaylist = await getPlaylist(elements);
+      setPlaylist(newPlaylist);
       setResetPreview(false);
-      getPlaylist(elements);
+    };
+
+    if (resetPreview && elements && elements.length > 0) {
+      handleUpdatePreview();
```
Comment (emettely) [inline src/Components/PaperEditor/ProgrammeScriptContainer/index.js:162]: Revert these changes as they should be 
```
      setPlaylist(playlistItems);
    };

    if (resetPreview && elements && elements.length > 0) {
      setResetPreview(false);
      getPlaylist(elements);
```

File: src/Components/PaperEditor/ProgrammeScriptContainer/index.js [hunk 4]
```diff
@@ -189,20 +206,18 @@ const ProgrammeScriptContainer = (props) => {
...
-      // either newText is empty or they hit cancel
-      console.log('Not editing');
     }
   };
 
   const onSortEnd = ({ oldIndex, newIndex }) => {
     const newElements = arrayMove(elements, oldIndex, newIndex);
+    console.log('handling reorder...');
+    handleReorder(newElements, oldIndex, newIndex);
     setElements(newElements);
```
Comment (emettely) [inline src/Components/PaperEditor/ProgrammeScriptContainer/index.js:221]:
```suggestion
    const updatedWords = await updateWordTimings(newElements, oldIndex, newIndex); 
    setElements(updatedWords);
```

File: src/Components/PaperEditor/ProgrammeScriptContainer/index.js [hunk 5]
```diff
@@ -166,14 +177,20 @@ const ProgrammeScriptContainer = (props) => {
     };
   });
 
+  const handleReorder = async (newElements, oldIndex, newIndex) => {
+    console.log('reorder elements: ', newElements);
+    const updatedWords = await updateWordTimings(newElements, oldIndex, newIndex);
+    setElements(updatedWords);
+    setResetPreview(true);
+  };
+
```
Comment (emettely) [inline src/Components/PaperEditor/ProgrammeScriptContainer/index.js:186]:
```suggestion
```

See simplification below in `onSortEnd`

File: src/Components/PaperEditor/ProgrammeScriptContainer/index.js [hunk 6]
```diff
@@ -215,7 +230,23 @@ const ProgrammeScriptContainer = (props) => {
     return elements.indexOf(insertElement);
   };
 
-  // // [old comment] TODO: needs to handle when selection spans across multiple paragraphs
```
Comment (emettely) [inline src/Components/PaperEditor/ProgrammeScriptContainer/index.js:218]: Maybe we shouldn't get rid of this comment?

Comment (allishultes) [reply src/Components/PaperEditor/ProgrammeScriptContainer/index.js:218]: I'll open a new ticket - probs best to have TODOs on the board v in the comments

**Comment by bennuttall** [comment]
Is double clicking a word in the Program Script supposed to work? I see a console log saying "handling double click" but nothing happens

Works well in the Transcript pane though 👍

File: src/Components/PaperEditor/ProgrammeScriptContainer/index.js [hunk 7]
```diff
@@ -134,7 +139,7 @@ const ProgrammeScriptContainer = (props) => {
         { startTime: 0, playlist: [] }
       );
 
-      const { playlist: playlistItems } = results;
```
Comment (emettely) [inline src/Components/PaperEditor/ProgrammeScriptContainer/index.js:137]: :D

File: src/Components/PaperEditor/ProgrammeScriptContainer/reset-word-timings.js [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+  return reindexedList;
+};
+
+const updateWordTimings = (elements) => {
+  const reindexedList = reindexList(elements);
+  const updatedElementTimings = reindexedList.reduce((result, element) => {
+
+    // Re-calcultate word timings
+    if (element.type === 'paper-cut') {
+      const paperCut = element;
```
Comment (emettely) [inline src/Components/PaperEditor/ProgrammeScriptContainer/reset-word-timings.js:18]: probably mutates the original element 🤔 would need to think of a way to deep clone the element
